### ISSUE TITLE
Closure of ldc-built programs is too large

### ISSUE THREAD

**Issue by CyberShadow** [NixOS/nixpkgs#324306]
### Describe the bug
```nix
{ lib, writeText, stdenv, ldc }:
stdenv.mkDerivation {
  name = "ldc-test";
  src = writeText "test.d" ''
    module test;
    import std.stdio;
    void main() {
      writeln("Hello, world!");
    }
  '';
  dontUnpack = true;
  buildInputs = [ ldc ];
  buildPhase = ''
    ldc2 -of=test $src
  '';
  installPhase = ''
    mkdir -p $out/bin
    mv test $out/bin
  '';
}
```

Building and getting the closure size gives us a number around 1.3 GiB:

```console
$ nix --extra-experimental-features nix-command path-info --json -S $(nix-build --no-out-link -E '(import <nixpkgs> {}).callPackage ./package.nix {}') | jq '.[0].closureSize' 
1395772728
```

Graph:

![out](https://github.com/NixOS/nixpkgs/assets/160894/3dc64c6c-b831-490a-baea-7246d1f4d3d1)

It looks like there are some things there which might not need to be there, such as:
- the bootstrap LDC
- LLVM
- considering that the test program was compiled with neither debug information nor is dynamically linked to Phobos, LDC itself

### Notify maintainers

CC @ThomasMader @lionello @jtbx 

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.34, NixOS, 24.11 (Vicuna), 24.11pre641786.d603719ec6e2`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/ksj02zhy7qc4dmqi8g2ga8p9hj5173dv-nixpkgs-patched`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by lionello** [comment]
Also see #111525

**Comment by TomaSajt** [comment]
@jtbx if you have the time could you do something about the fact that `ldc` has references to `ldc-bootstrap`?

**Comment by jtbx** [comment]
Yes! #393645

**Comment by TomaSajt** [comment]
I saw that https://github.com/NixOS/nixpkgs/pull/324400 was merged which took care of most of the issue.

@CyberShadow are there any more cases where `ldc` is bloating the closure size?

---

Also: should we actually patch the compiler so that any `/nix/store` prefixes embedded into the file via `assert` or `__FILE__` get replaced with something else?
like with `--remap-path-prefix` in `rustc`, see https://github.com/ldc-developers/ldc/issues/3908

**Comment by CyberShadow** [comment]
> [@CyberShadow](https://github.com/CyberShadow) are there any more cases where `ldc` is bloating the closure size?

Yes, I don't remember all of the details, but that fixed only one of the specific cases.

> Also: should we actually patch the compiler so that any `/nix/store` prefixes embedded into the file via `assert` or `__FILE__` get replaced with something else?

IMHO, this is fine as long as it is not completely useless (e.g. it aids debugging) and it does not cause the closure to be unjustifiably big. #324000 put the referenced standard library source code in its own derivation output, so the paths are still valid, but don't pull in the rest of the compiler.

I opened https://github.com/ldc-developers/ldc/issues/4695 upstream for the shared library counterpart, in case someone experienced with CMake wants to have a look.
